### ISSUE TITLE
Add missing fields into Tx.

### ISSUE THREAD

**Issue by zmrocze** [input-output-hk/plutus-apps#471]
### Describe the feature you'd like

Currently the `Tx` type taken from `plutus-ledger-api` lacks:
 - [ ] metadata
 - [ ] withdrawals
 - [ ] certificates
 
These should be added together with updating:
  - `toCardanoTxBody` translation
  - validation in emulator
  - adding related constraints in `plutus-ledger-constraints`
  - ...

**Comment by zmrocze** [comment]
I've added this issue as I lacked a place discussing changing the `Tx` and the old one taken as is from plutus-ledger-api is as I understand only temporary.

related issues:
 - metadata #93
  - withdrawals #395 
  - Should witnesses be stored on the side? #459

**Comment by zmrocze** [comment]
I went with such a change: [Ledger/Tx/Internal/Tx.hs](https://github.com/zmrocze/plutus-apps/blob/e1248a02dc87b799aeba0869043fe27e9a3936ed/plutus-ledger/src/Ledger/Tx/Internal.hs#L78) here #468.

Making these changes: 
 - moving out all script and datum witnesses into seperate txScripts and txData fields, leaving hashes in inputs/etc. Before these were attached to inputs, but in a Maybe field, so this doesn't change that expected witnesses might or might not be present.
 - attaching redeemers to inputs/mints (conversely ^). Before the `Redeemers` map was very loosely related to scripts with `RedeemerPtr`'s, which makes it hard to manually construct transactions with more than 1 mint script. I took suggestion from TxConstraints where constraints contain redeemers but refer to scripts only by hashes. 
 - adding certificates/withdrawals/metadata fields
 
 I mostly fixed compiling and added withdrawals translation in `toCardanoTxBody` based on [plutus-simple-model](https://github.com/mlabs-haskell/plutus-simple-model/blob/main/src/Plutus/Test/Model/Fork/CardanoAPI.hs).
 
Adding validation for withdrawals/certificates is harder as `ValidationContext` doesn't contain state needed to validate these (like rewards/delegations/registered credentials).

I would enjoy greatly discussing the expected shape for `Tx` and this proposition for it.
